[Experimental] Product Filters Redesign > Filters: Implement Dynamic Block Title and Icon

thealexandrelara opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #48637 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Navigate to Product Catalog template (Appearance > Editor > Templates > Product Catalog template - or any archive template).
2. Click on "Edit" to open the WordPress editor.
3. In the editor, you should see a block inserter icon or an option to "Add Block". Click on it.
4. Search for "Product Filters (Experimental)" in the block inserter or browse the available blocks until you find them. Add it to your content.
5. Open the List View and make sure the Product Filters (Experimental) contains the correct inner blocks: Filters, Active (Experimental), Price (Experimental), Status (Experimental), Color (Experimental), and Rating (Experimental). 
6. Make sure that each inner block contain their own icon according to the design.

The "Color (Experimental)" block should show if you're using the default WooCommerce sample data products. If you have more products, the block name should be the attribute with the highest product count on your store.

| Before | Current |
|--------|--------|
| ![CleanShot 2024-06-19 at 14 39 57@2x](https://github.com/woocommerce/woocommerce/assets/20469356/fb5ce7e0-e98a-4476-a1a1-97dd86e2aabf) | ![CleanShot 2024-06-21 at 17 53 31@2x](https://github.com/woocommerce/woocommerce/assets/20469356/efe7743e-2783-4b50-acad-1f88569d7f23) | 

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead --> This PR does not require a changelog as the changes are behind an experimental flag and will not affect users.

</details>

thealexandrelara committed:
Modify Product Filters block to use block variations

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

thealexandrelara committed:
Add dynamic blocks icon and title

thealexandrelara committed:
Make Attribute Filter block title dynamic

thealexandrelara committed:
Merge branch 'trunk' into feat/48637-implement-dynamic-block-icon-and-title

thealexandrelara committed:
Add E2E tests for the Product Filters block

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filter/types.ts:
```diff
@@ -13,6 +13,7 @@ export type FilterType = keyof typeof BLOCK_NAME_MAP;
 export type BlockAttributes = {
 	filterType: FilterType;
 	heading: string;
+	providerNameSlug: string;
```
Can you explain why we don't use the existing attribute (`filterType`) and add this new one?

thealexandrelara replied:
I just removed the providerNameSlug, your comment made me realize it is not necessary (ref: 7d557851dc2a129a1f941a1f6221f48c1397b030).

dinhtungdu reviewed (commented):
I can confirm this PR is working as indended following the testing instructions. I left a comment related to the newly added attribute, otherwise, this is LGTM.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filter/block-variations.tsx:
```diff
@@ -0,0 +1,124 @@
+ * `isActive` function is used to find a variation match from a created
+ *  Block by providing its attributes.
+ */
+( variations as Variation[] ).forEach( ( variation ) => {
+	variation.isActive = (
+		blockAttributes: BlockAttributes,
+		variationAttributes: BlockAttributes
+	) =>
+		blockAttributes.providerNameSlug ===
+		variationAttributes.providerNameSlug;
+} );
```
Isn't this equal to `isActive: [ 'providerNameSlug' ]`

thealexandrelara replied:
TBH this is the first time I'm working with block variations, so I was just following the instructions from the docs. Thank you for help me realize this is not needed. I just pushed a commit with your suggestion in 7d557851dc2a129a1f941a1f6221f48c1397b030.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filter/block-variations.tsx:
```diff
@@ -0,0 +1,124 @@
+};
+
+/**
+ * Add `isActive` function to all `embed` variations, if not defined.
```
The comment mentions the Embed block.

thealexandrelara replied:
I didn't understand what you mean by "Embed" block. This comment is just to explain that that piece of the code is responsible for adding the `isActive` function to all the embed block variations.

dinhtungdu replied:
I said that because this comment is the same as [one](https://github.com/WordPress/gutenberg/blob/f4f770d74e794bdc586a6f0919ecc4eb20d1003d/packages/block-library/src/embed/variations.js#L366-L370) of Embed block.

My interpretation of this comment is we're adding isActive to all variations of the Embed block, note that the word embed is put inside backticks. 

> if not defined.

Currently, the following function always adds `isActive` to the variation, regardless of whether the variation defines it or not.

thealexandrelara replied:
Thank you for pointing this out. I adjusted the text in: https://github.com/woocommerce/woocommerce/pull/48638/commits/0314e6cdd41c70b081d95538ef746cfedd4c0cd2

thealexandrelara committed:
Remove unnecessary block attribute

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filter/block-variations.tsx:
```diff
@@ -0,0 +1,116 @@
+				variationAttributes: BlockAttributes
+		  ) => boolean )
+		| string[];
+};
```
What do you think about using the `BlockVariation` [type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1c20553fd9da69247ff3ffe1206be8222439615a/types/wordpress__blocks/index.d.ts#L661-L679) here?

thealexandrelara replied:
Awesome suggestion! I just replaced the custom type by the BlockVariation type from the @wordpress/blocks library. 9fc4c16494838c3ef5f2df0924aab254799cbc39

dinhtungdu reviewed (commented):
The PR is still working as expected following the testing instructions. I can confirm the titles and icons of filter blocks are correct, with an exception of the attribute filter block.

Adding the Product Filters block to the template, I can see the attribute with most terms are automatically selected, and the block title is set to the attribute title correctly. But if I change the attribute, the title remains unchanged. See the screencast: 

https://github.com/woocommerce/woocommerce/assets/5423135/a540fa85-f760-4d08-b4d7-7d913044ece7

roykho commented:
A couple of observations. I am finding that the filter icons are different when you drill down into the inner blocks. Would it be more clear if the icons matched?

![Google Chrome - Product Catalog ‹ Template ‹ woo-blocks-dev ‹ Editor — WordPress 2024-06-26 at 6 30 58 AM](https://github.com/woocommerce/woocommerce/assets/2132595/5339f8fc-2470-4222-8c3f-b53041bd8bae)

I understand the attributes will show for the ones with most product count but do you think it could be confusing? I don't have strong opinion here but maybe we could use the attribute title but also what the filter is for? For example Color (attributes).

If the user removes a filter by accident, how do they re-add it?

Changing the attributes for example from Color to Size, the titles did not change.
![Google Chrome - Product Catalog ‹ Template ‹ woo-blocks-dev ‹ Editor — WordPress 2024-06-26 at 6 36 29 AM](https://github.com/woocommerce/woocommerce/assets/2132595/38907b4e-9b8a-4a02-8fe1-5a79f428e542)

thealexandrelara committed:
Improve clarity of the comment text

thealexandrelara committed:
Add new icons for individual filters

thealexandrelara committed:
Dynamically set the title and heading of the block based on the selected attribute

thealexandrelara committed:
Add E2E test to check for block title and heading change on selecting an attribute

thealexandrelara committed:
Replace custom block variation type by BlockVariation type from @wordpress/blocks

thealexandrelara commented:
> Adding the Product Filters block to the template, I can see the attribute with most terms are automatically selected, and the block title is set to the attribute title correctly. But if I change the attribute, the title remains unchanged.

Oh, great catch! I just fixed this in https://github.com/woocommerce/woocommerce/pull/48638/commits/2dde312eecd0bd47bbc1081a2df4bcc22b71172b. When merchants choose a different attribute; the block title and heading content are automatically updated.

> I am finding that the filter icons are different when you drill down into the inner blocks. Would it be more clear if the icons matched?

I changed the icons from each filter block to match the wrapper block (https://github.com/woocommerce/woocommerce/pull/48638/commits/ffa02b3b0b6bbc0eb4462e271247f090f22c415b); however, in the future, we will probably replace the name and icon of each filter block (in the design, they are called Filter Options).

> Changing the attributes for example from Color to Size, the titles did not change.

Thank you for raising this point. This is now fixed in https://github.com/woocommerce/woocommerce/pull/48638/commits/2dde312eecd0bd47bbc1081a2df4bcc22b71172b.

> Also how does user want to have more than on attribute filter block. However, please keep in mind that inserter behavior is out of this PR scope.

Yeah, we will have to discuss this further in the future and make decisions regarding block insertions. Roy worked on an issue related to this (https://github.com/woocommerce/woocommerce/issues/46981). Basically, we will have to work on ways to limit block insertions for some filters while allowing other filters to be freely added or removed.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filter/block-variations.tsx:
```diff
@@ -0,0 +1,105 @@
+ *  Block by providing its attributes.
+ */
+variations.forEach( ( variation ) => {
+	variation.isActive = () => [ 'filterType' ];
```
This should be `string[]` or `() => boolean`, not `() => string[]`.
```suggestion
	variation.isActive = [ 'filterType' ];
```

The type is actually incorrect, we should fire a pull to fix it.
```
Diagnostics:
1. Type 'string[]' is not assignable to type '(blockAttributes: BlockAttributes, variationAttributes: BlockAttributes) => boolean | string[]'.
     Type 'string[]' provides no match for the signature '(blockAttributes: BlockAttributes, variationAttributes: BlockAttributes): boolean | string[]'. [2322]
```

thealexandrelara replied:
Fixed in b664a7b74f0ab496820b2e0601221eb07eab982b. From the type definition this is an acceptable way, but I changed to be an array of strings.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/edit.tsx:
```diff
@@ -30,36 +30,36 @@ const TEMPLATE: InnerBlockTemplate[] = [
 		{
 			filterType: 'price-filter',
-			heading: __( 'Price', 'woocommerce' ),
+			heading: 'Filter by Price',
```
Can you explain why do we change this? The current value matches the new design. Also, this heading, active and stock ones miss the translation wrapper function.

thealexandrelara replied:
Probably copy-pasted from one of the individual blocks, but I just fixed in https://github.com/woocommerce/woocommerce/pull/48638/commits/d9a31412f4ecdefa022d9d2144a53eee94555ad6

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filter/inner-blocks/attribute-filter/edit.tsx:
```diff
@@ -138,6 +142,47 @@ const Edit = ( props: EditProps ) => {
+				const parentBlockId = parentBlocksByBlockName[ 0 ];
+
+				const parentBlock = getBlock( parentBlockId );
+				const headerGroupBlock = parentBlock?.innerBlocks.find(
+					( block ) => block.name === 'core/group'
+				);
+				const headingBlock = headerGroupBlock?.innerBlocks.find(
+					( block ) => block.name === 'core/heading'
+				);
```
If the header group block is wrapped in another group for layout purpose, this is likely to be failed to get the heading. I think we should recursively find the first heading block instead of relying on a fixed structure.

thealexandrelara replied:
It would also fail if someone added another Heading block before the current heading. The only way to always find the correct block would be if the core/heading block accepted some custom attribute or way to identify it as the “default” block heading. Do you know if this is possible?

It’s impossible to cover all possibilities, so we should focus on providing the best experience for the way the design was initially intended. Users are welcome to modify and customize as they want; however, it comes at a cost where some functionalities might not work the same as when the block is used without modifications to the layout structure, etc.

dinhtungdu replied:
> It would also fail if someone added another Heading block before the current heading

I agree that we shouldn't touch user created content, so let's leave it as it is now 👍🏽

dinhtungdu reviewed (commented):
@thealexandrelara thanks for the updates! I found some issues with the latest commits, both in my local instance and test JN site. The screenshot below was taken on a JN site.

- The title and icon of the Product Filter block variations are always set to Active one.
- The heading of Attribute filter set to undefined if there is something wrong happened.
- The inner filter block headings are incorrect comparing to the design.

![image](https://github.com/woocommerce/woocommerce/assets/5423135/6b37f0b2-811d-4860-8e6c-8da954c56e11)

> I changed the icons from each filter block to match the wrapper block (https://github.com/woocommerce/woocommerce/commit/ffa02b3b0b6bbc0eb4462e271247f090f22c415b); however, in the future, we will probably replace the name and icon of each filter block (in the design, they are called Filter Options).

I wonder why we don't use the Filter Options icon from the design right away?

thealexandrelara committed:
Replace variation `isActive` function by an array of strings

thealexandrelara committed:
Fix heading for each individual filter

thealexandrelara committed:
Add a default attribute name when there is no attributes registered in the store

thealexandrelara commented:
Thank you for reviewing the PR again, @dinhtungdu . Great catches! I have addressed the issues you raised.

>  The title and icon of the Product Filter block variations are always set to Active one.

Could you try it again? For some reason Github CI didn't update this PR with its latest version. I got the same issue when I tried running on WordPress Playground and/or JN site, but now it seems to be fixed after re-running the CI.

> The heading of Attribute filter set to undefined if there is something wrong happened.

Fixed in ffa347c2f73ab019541b73b99be0d909ebe51bbd. A default "Attribute" name is set to the block title and heading when no attribute is found in the store.

> The inner filter block headings are incorrect comparing to the design.

Fixed in d9a31412f4ecdefa022d9d2144a53eee94555ad6

roykho reviewed (commented):
Thanks for all the updates/fixes. This is looking good and testing well. Just one note is the icons. I am not sure if we need/should move them to the @woocommerce/icons package in case it needs to be reused? Other than that, I am pre-approving.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/edit.tsx:
```diff
@@ -30,21 +30,21 @@ const TEMPLATE: InnerBlockTemplate[] = [
 		{
 			filterType: 'stock-filter',
-			heading: __( 'Status', 'woocommerce' ),
+			heading: 'Status',
```
These strings are missing i18n wrapper.

thealexandrelara replied:
Fixed in ce5a2af9613261a8f970d4d2b3d371d6a99ad6e4. Thank you for reporting these!

dinhtungdu commented on plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts:
```diff
@@ -62,26 +62,24 @@ test.describe( `${ blockData.name }`, () => {
 
-		const priceHeading = block.getByRole( 'document', {
-			name: 'Price',
+		const priceHeading = block.getByText( 'Filter by Price', {
```
This should be Price right, as we don't use old titles anymore? The E2E test is falling by the way, I think it relates to this.

thealexandrelara replied:
You're right; I think my brain melted down on Friday 😅. Fixed in ce5a2af9613261a8f970d4d2b3d371d6a99ad6e4

dinhtungdu reviewed (commented):
This PR is testing great on my end. Thank you for addressing all feedback! I also agree with you about working on Filter Options in another PR.

Beside the review comments, I agree with as @roykho on extracting icon to https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/assets/js/icons.

Otherwise, this is LGTM.

thealexandrelara committed:
Fix i18n strings

thealexandrelara committed:
Merge branch 'trunk' into feat/48637-implement-dynamic-block-icon-and-title

thealexandrelara commented:
Thank you all for reviewing this and reporting these issues. I’ve addressed all the points raised. Regarding the icons, I just created an [issue](https://github.com/woocommerce/woocommerce/issues/48990) for that, so I can work on it soon and we can unblock this PR to be merged.
